handlers/clickatell/clickatell.go

```diff
@@ -1,5 +1,269 @@
+		text, _ = new(mime.WordDecoder).DecodeHeader(text)
+	}
+
+	// build our msg
+	msg := h.Backend().NewIncomingMsg(channel, urn, utils.CleanString(text)).WithReceivedOn(date.UTC()).WithExternalID(ctIncomingMessage.SmsID)
+
+	// and write it
+	err = h.Backend().WriteMsg(ctx, msg)
+	if err != nil {
+		return nil, err
+	}
+
+	return []courier.Event{msg}, courier.WriteMsgSuccess(ctx, w, r, []courier.Msg{msg})
+}
+
+func decodeUTF16BE(b []byte) (string, error) {
+	if len(b)%2 != 0 {
+		return "", fmt.Errorf("Must have even length byte slice")
+	}
+	u16s := make([]uint16, 1)
+	ret := &bytes.Buffer{}
+	b8buf := make([]byte, 4)
+
+	lb := len(b)
+	for i := 0; i < lb; i += 2 {
+		u16s[0] = uint16(b[i+1]) + (uint16(b[i]) << 8)
+		r := utf16.Decode(u16s)
+		n := utf8.EncodeRune(b8buf, r[0])
+		ret.Write(b8buf[:n])
+	}
+	return ret.String(), nil
+}
+
+// SendMsg sends the passed in message, returning any error
+func (h *handler) SendMsg(ctx context.Context, msg courier.Msg) (courier.MsgStatus, error) {
+	username := msg.Channel().StringConfigForKey(courier.ConfigUsername, "")
+	if username == "" {
+		return nil, fmt.Errorf("no username set for CT channel")
+	}
+
+	password := msg.Channel().StringConfigForKey(courier.ConfigPassword, "")
+	if password == "" {
+		return nil, fmt.Errorf("no password set for CT channel")
+	}
+
+	apiID := msg.Channel().StringConfigForKey(courier.ConfigAPIID, "")
+	if apiID == "" {
+		return nil, fmt.Errorf("no api_id set for CT channel")
+	}
+
+	unicodeSwitch := "0"
+	text := courier.GetTextAndAttachments(msg)
+	if !gsm7.IsGSM7(text) {
+		replaced := gsm7.ReplaceNonGSM7Chars(text)
+		if gsm7.IsGSM7(replaced) {
+			text = replaced
+		} else {
+			unicodeSwitch = "1"
+		}
+	}
+
+	re := regexp.MustCompile(`^ID: (.*)`)
+
+	status := h.Backend().NewMsgStatusForID(msg.Channel(), msg.ID(), courier.MsgErrored)
+	parts := handlers.SplitMsg(text, maxMsgLength)
+	for _, part := range parts {
+		form := url.Values{
+			"api_id":   []string{apiID},
+			"user":     []string{username},
+			"password": []string{password},
+			"from":     []string{strings.TrimPrefix(msg.Channel().Address(), "+")},
+			"concat":   []string{"3"},
+			"callback": []string{"7"},
+			"mo":       []string{"1"},
+			"unicode":  []string{unicodeSwitch},
+			"to":       []string{strings.TrimPrefix(msg.URN().Path(), "+")},
+			"text":     []string{part},
+		}
+
+		encodedForm := form.Encode()
+		partSendURL := fmt.Sprintf("%s?%s", sendURL, encodedForm)
+
+		req, err := http.NewRequest(http.MethodGet, partSendURL, nil)
+		req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+		rr, err := utils.MakeHTTPRequest(req)
+
+		// record our status and log
+		log := courier.NewChannelLogFromRR("Message Sent", msg.Channel(), msg.ID(), rr)
+		status.AddLog(log)
+		if err != nil {
+			log.WithError("Message Send Error", err)
+			return status, nil
+		}
+
+		if rr.StatusCode != 200 && rr.StatusCode != 201 && rr.StatusCode != 202 {
```

Comment:
These types of errors are automatically caught by `MakeHTTPRequest` (it tests for a 2** response and throws an error if not)